### PR TITLE
Fix Refit Cache for mamba

### PR DESCRIPTION
**Pull request by wdykas**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
The problem is that in non colocated inference in Nemo rl the _A_neg_exp_cache wasn't getting updated so it was stale. This adds  the idea of post hooks to fix the issue.

### CODE REVIEW

**Comment by wdykas** [comment]
/ok to test [7d6bbb1](https://github.com/NVIDIA/Megatron-LM/pull/6672/commits/7d6bbb12ed0958471f6a3e9e91c5dd59fe7dd340)

**Comment by wdykas** [comment]
/ok to test [38e05ee](https://github.com/NVIDIA/Megatron-LM/commit/38e05ee1ecfe6628716e3d4a6e775a50c67491ac)

**Review by santhnm2 — APPROVED** [review]

**Review by sidsingh-nvidia — APPROVED** [review]

**Review by shanmugamr1992 — APPROVED** [review]

File: tests/unit_tests/ssm/test_mamba_mixer.py [hunk 1]
```diff
@@ -107,6 +107,31 @@ def test_variable_batch_size_inference(self):
...
+        cache_ptr = mixer._A_neg_exp_cache.data_ptr()
+
+        with torch.no_grad():
+            mixer.A_log.add_(1.0)
+        mixer.train()
+        mixer.eval()
+
+        torch.testing.assert_close(mixer._A_neg_exp_cache, -torch.exp(mixer.A_log.float()))
+        assert mixer._A_neg_exp_cache.data_ptr() == cache_ptr
+        assert not mixer._A_neg_exp_cache_stale
```
Comment (tdene) [inline tests/unit_tests/ssm/test_mamba_mixer.py]: These tests are identical; can we combine it into a single parameterized test?

File: megatron/core/resharding/refit.py [hunk 2]
```diff
@@ -490,3 +500,4 @@ def reshard_model_weights(
     execute_reshard_plan(
         plan, src_core, tgt_core, service=service, group=group, transform=transform
     )
+    _run_post_refit_hooks(tgt_core)
```
Comment (tdene) [inline megatron/core/resharding/refit.py]: Why not put this deeper, inside `execute_reshard_plan`? No one is calling `execute_reshard_plan` standalone today, but just for code cleanliness.

File: megatron/core/ssm/mamba_mixer.py [hunk 3]
```diff
@@ -1022,6 +1022,19 @@ def ssm_prefill(
...
+        self._A_neg_exp_cache.copy_(-torch.exp(self.A_log.float()))
+        self._A_neg_exp_cache_stale = False
+
+    def post_refit(self) -> None:
+        """Refresh the decode cache after refit updates ``A_log`` in place.
+
+        Keeping the cache tensor itself preserves addresses captured by CUDA graphs.
+        """
+        self._refresh_A_neg_exp_cache()
+
```
Comment (tdene) [inline megatron/core/ssm/mamba_mixer.py]: Why have two separate methods? Seems like `refresh_cache` should just be a public method so we can get rid of `post_refit`.

File: megatron/core/ssm/mamba_mixer.py [hunk 4]
```diff
@@ -1049,9 +1060,10 @@ def _get_batch_invariant_decoder(self) -> MambaBatchInvariantDecode:
     def train(self, mode: bool = True):
         """Mark the decode cache stale; weights may have updated."""
```
Comment (tdene) [inline megatron/core/ssm/mamba_mixer.py]: This docstring is stale now.

File: megatron/core/ssm/mamba_mixer.py [hunk 5]
```diff
@@ -1049,9 +1060,10 @@ def _get_batch_invariant_decoder(self) -> MambaBatchInvariantDecode:
...
         if mode:
-            # only mark stale when switching to training mode.
-            # otherwise retain the staleness state.
             self._A_neg_exp_cache_stale = True
+        elif self._A_neg_exp_cache_stale:
+            # CUDA graph replay bypasses the Python lazy-refresh path.
+            self._refresh_A_neg_exp_cache()
```
Comment (tdene) [inline megatron/core/ssm/mamba_mixer.py]: This makes it so that running `.eval()` will always clear the flag and refresh the cache.

There's some edge conditions possible here, like running `.eval()` before initializing the model weights. It's highly unlikely this would ever happen, I don't think it should be fixed, just pointing it out.

File: tests/unit_tests/ssm/test_mamba_mixer.py (hunk 1 above)
Comment (wdykas) [reply tests/unit_tests/ssm/test_mamba_mixer.py]: address

File: megatron/core/resharding/refit.py (hunk 2 above)
Comment (wdykas) [reply megatron/core/resharding/refit.py]: adress

File: megatron/core/ssm/mamba_mixer.py (hunk 3 above)
Comment (wdykas) [reply megatron/core/ssm/mamba_mixer.py]: adress

**Review by tdene — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Comment by wdykas** [comment]
/ok to test [f6ca8d5](https://github.com/NVIDIA/Megatron-LM/pull/6672/commits/f6ca8d512de389a9199967c482d7dcd4ea88a401)
